docs/DesignDocs/AttestationAPIsWithFlexibleClaims.md

```diff
@@ -0,0 +1,476 @@
+   * generates a self-signed x.509 certificate with embedded evidence generated by
+   * an attester plugin for the enclave, but it also allows a user to pass in
+   * runtime and inittime claims.
+   *
+   * @experimental
+   *
+   * @param[in] format_id The format id of the evidence to be generated.
+   *
+   * @param[in] subject_name a string containing an X.509 distinguished
+   * name (DN) for customizing the generated certificate. This name is also used
+   * as the issuer name because this is a self-signed certificate
+   * See RFC5280 (https://tools.ietf.org/html/rfc5280) for details
+   * Example value "CN=Open Enclave SDK,O=OESDK TLS,C=US"
+   *
+   * @param[in] private_key a private key used to sign this certificate
+   * @param[in] private_key_size The size of the private_key buffer
+   * @param[in] public_key a public key used as the certificate's subject key
+   * @param[in] public_key_size The size of the public_key buffer.
+   * @param[in] inittime_claims an optional flat buffer representing the inittime
+   * claims.
+   * @param[in] inittime_claims_size The size of the inittime_claims in bytes.
+   * @param[in] runtime_claims an optional flat buffer representing the runtime
+   * claims.
+   * @param[in] runtime_claims_size The size of the runtime_claims in bytes.
+   * @param[in] optional_parameters The optional format-specific input parameters.
+   * @param[in] optional_parameters_size The size of optional_parameters in bytes.
+   *
+   * @param[out] output_certificate a pointer to buffer pointer
+   * @param[out] output_certificate_size size of the buffer above
+   *
+   * @return OE_OK on success
+   */
+  oe_result_t oe_get_attestation_certificate_with_evidence_v3(
+      const oe_uuid_t* format_id,
+      const unsigned char* subject_name,
+      uint8_t* private_key,
+      size_t private_key_size,
+      uint8_t* public_key,
+      size_t public_key_size,
+      uint8_t* runtime_claims,
+      size_t runtime_claims_size,
+      uint8_t* inittime_claims,
+      size_t inittime_claims_size,
+      const void* optional_parameters,
+      size_t optional_parameters_size,
+      uint8_t** output_certificate,
+      size_t* output_certificate_size);
+
+  ```
+  The API introduces four arguments that allow developers to optionally
+  pass in run- and init-time claims.
+  - `runtime_claims`
+  - `runtime_claims_size`
+  - `inittime_claims`
+  - `inittime_claims_size`
+
+- oe_verify_attestation_certificate_with_evidence_v3
+  ```c
+  /**
+   * oe_verify_attestation_certificate_with_evidence_v3
+   *
+   * This function is similar to
+   * oe_verify_attestation_certificate_with_evidence_v2 that has the same
+   * prototype but with the ability to verify a ceritifacte generated by
+   * oe_get_attestation_certificate_with_evidence_v3.
+   *
+   * @experimental
+   *
+   * @param[in] certificate_in_der A pointer to buffer holding certificate
+   * contents in DER format.
```

Comment:
question: I know this is experimental, but do we really want separate APIs for certificates vs (say) JWTs vs CWTs?
oe_verify_evidence in contrast has a format_id to allow flexibility of format.   I might ask the same about the format of the endorsements_buffer.

Reply:
This sounds like supporting a format of "certificate with embedded evidence" for the endorsement. In this case, the `evidence_buffer` can be NULL.

```
        result = oe_verify_evidence(
            NULL,
            NULL,
            0,
            endorsements_buffer, /* include the certificate with the embedded evidence */
            endorsements_buffer_size,
            policies,
            policies_size,
            claims,
            claims_length);
``` 

But this requires significant changes to the  `oe_get_evidence` API to support this proposal unless we still keep the `oe_get_attestation_certificate_with_evidence_v3` but drop `oe_verify_attestation_certificate_with_evidence_v3`.